otx/algorithms/detection/adapters/mmdet/hooks/det_class_probability_map_hook.py

```diff
@@ -121,3 +124,127 @@ def forward_single(x, cls_convs, conv_cls):
+        img_metas *= batch_size
+        proposals = self._module.rpn_head.simple_test_rpn(x, img_metas)
+        test_cfg = copy.deepcopy(self._module.roi_head.test_cfg)
+        test_cfg["max_per_img"] = self._max_detections_per_img
+        test_cfg["nms"]["iou_threshold"] = 1
+        test_cfg["nms"]["max_num"] = self._max_detections_per_img
+        det_bboxes, det_labels = self._module.roi_head.simple_test_bboxes(
+            x, img_metas, proposals, test_cfg, rescale=False
+        )
+        return det_bboxes, det_labels
+
+    def _get_saliency_maps_from_mask_predictions(
+        self, x: torch.Tensor, det_bboxes: List[torch.Tensor], det_labels: List[torch.Tensor]
+    ) -> List[List[Optional[np.ndarray]]]:
+        _bboxes = [det_bboxes[i][:, :4] for i in range(len(det_bboxes))]
+        mask_rois = bbox2roi(_bboxes)
+        mask_results = self._module.roi_head._mask_forward(x, mask_rois)
+        mask_pred = mask_results["mask_pred"]
+        num_mask_roi_per_img = [len(det_bbox) for det_bbox in det_bboxes]
+        mask_preds = mask_pred.split(num_mask_roi_per_img, 0)
+
+        batch_size = x[0].shape[0]
+
+        scale_x = self._input_img_shape[1] / self._saliency_map_shape[1]
+        scale_y = self._input_img_shape[0] / self._saliency_map_shape[0]
+        scale_factor = torch.FloatTensor((scale_x, scale_y, scale_x, scale_y))
+        test_cfg = self._module.roi_head.test_cfg.copy()
+        test_cfg["mask_thr_binary"] = -1
+
+        saliency_maps = []  # type: List[List[Optional[np.ndarray]]]
+        for i in range(batch_size):
+            saliency_maps.append([])
+            for j in range(self._module.roi_head.mask_head.num_classes):
+                saliency_maps[i].append(None)
+
+        for i in range(batch_size):
+            if det_bboxes[i].shape[0] == 0:
+                continue
+            else:
+                segm_result = self._module.roi_head.mask_head.get_seg_masks(
+                    mask_preds[i],
+                    _bboxes[i],
+                    det_labels[i],
+                    test_cfg,
+                    self._saliency_map_shape,
+                    scale_factor=scale_factor,
+                    rescale=True,
+                )
+                for class_id, segm_res in enumerate(segm_result):
+                    if segm_res:
+                        saliency_maps[i][class_id] = np.mean(np.array(segm_res), axis=0)
+        return saliency_maps
+
+    @staticmethod
+    def _normalize(saliency_maps: List[List[Optional[np.ndarray]]]) -> List[List[Optional[np.ndarray]]]:
+        batch_size = len(saliency_maps)
+        num_classes = len(saliency_maps[0])
+        for i in range(batch_size):
+            for class_id in range(num_classes):
+                per_class_map = saliency_maps[i][class_id]
+                if per_class_map is not None:
+                    max_values = np.max(per_class_map)
+                    per_class_map = 255 * (per_class_map) / (max_values + 1e-12)
```

Comment:
`per_class_map = None` covers the cases, when we cannot generate any saliency map for this particular target class. This can be the case for e.g. `MaskRCNNRecordingForwardHook`, because it actually generate sal maps from model predictions ( it is rare the case when all possible classes are getting predicted for each image)

Reply:
> we know that saliency map min_values are 0 in any case

This is not true in the general case. For example if you remove softmax activation from `DetClassProbabilityMapHook`, there will be negative values (because there can be negative values in the detection activations). So I would propose to keep it for the seek of generality, especially if we will unify the normalization for all methods.

Reply:
I'm sorry if I asked not really clearly, but it was a question: why don't we calculate and subtract min_values during normalization?

Reply:
`per_class_map = None covers the cases, when we cannot generate any saliency map for this particular target class.`
Got it, thank you. I thought, in that case, we'll get a saliency map with all zeros, so we don't need to make a special way to postprocess them

Reply:
> I'm sorry if I asked not really clearly, but it was a question: why don't we calculate and subtract min_values during normalization?

We should in general case - when/if we move normalization to the common part for all methods. In this particular example, for `MaskRCNNRecordingForwardHook` map combined from segmentation masks, which cannot be negative (min value is zero). So it did not subtracted min_values

Reply:
> I thought, in that case, we'll get a saliency map with all zeros, so we don't need to make a special way to postprocess them

This is an option. But the downside of it is memory overhead

Reply:
Thank you for explanation